### PR TITLE
feat: DAO SDK

### PR DESCRIPTION
**Pull request by n0izn0iz**
This implement a DAO SDK inspired by [DA0-DA0](https://github.com/DA0-DA0/dao-contracts)

The code in this PR has been extracted from https://github.com/gnolang/gno/pull/1154 and cleaned (reviews on the previous PR should have been addressed)

It includes:
- Interfaces for the DAO `gno.land/p/demo/dao_maker/dao_interfaces`
- A reference implementation for the core of the DAO `gno.land/p/demo/dao_maker/dao_core`
- A membership-based voting module `gno.land/p/demo/dao_maker/dao_voting_group`
- A single-choice proposal module `gno.land/p/demo/dao_maker/dao_proposal_single`
- An example DAO `gno.land/r/demo/dao_maker/dao_realm`
- A registry to dynamically interact with any registered DAO using it's address `gno.land/r/demo/dao_maker/dao_registry`

For usage see the [tutorial](https://github.com/TERITORI/gno/blob/1555f35be0b9ad8f65c2179d9a9b8c8f2a79aff4/docs/how-to-guides/creating-dao.md), it is integrated into docusaurus

Screenshot 2024-04-12 at 18 41 10

---

There is an example membership-based DAO at `gno.land/r/demo/teritori/dao_realm` that can administer itself and a grc20 token at `gno.land/r/demo/teritori/tori`

Here is it's GnoWeb Render:
Screenshot 2024-04-12 at 18 28 41

---

This powers the "Organizations" feature of Teritori dApp on Gno as shown below

The DAO creation UI uses a Gno template and instantiates a new realm.

Screenshot 2024-04-12 at 18 43 23
Screenshot 2024-04-12 at 18 44 50
Screenshot 2024-04-12 at 18 44 15
Screenshot 2024-04-12 at 18 42 26

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: examples/gno.land/p/demo/teritori/dao_core/dao_core.gno
Comment (notJoon) [inline examples/gno.land/p/demo/teritori/dao_core/dao_core.gno:141]: When concatenating strings, using `strings.Builder` instead of the `+=` operator can reduce unnecessary memory usage.

```suggestion
func (d *daoCore) Render(path string) string {
    var sb strings.Builder
    sb.WriteString("# DAO Core\n")

    votingInfo := d.votingModule.Info()
    sb.WriteString("## Voting Module: ")
    sb.WriteString(votingInfo.String())
    sb.WriteString("\n")

    sb.WriteString(d.votingModule.Render(""))

    sb.WriteString("## Supported Messages:\n")
    sb.WriteString(d.registry.Render())

    sb.WriteString("## Proposal Modules:\n")
    for i, propMod := range d.proposalModules {
        if !propMod.Enabled {
            continue
        }
        info := propMod.Module.Info()
        sb.WriteString("### #")
        sb.WriteString(strconv.Itoa(i))
        sb.WriteString(": ")
        sb.WriteString(info.String())
        sb.WriteString("\n")
        sb.WriteString(propMod.Module.Render(""))
    }

    sb.WriteString("\n")
    return sb.String()
}
```

File: examples/gno.land/p/demo/teritori/dao_interfaces/messages_registry.gno
Comment (notJoon) [inline examples/gno.land/p/demo/teritori/dao_interfaces/messages_registry.gno:64]: For the same reason, it would be good to use `strings.Builder` here as well.
```suggestion
func (r *MessagesRegistry) Render() string {
    var sb strings.Builder

    r.handlers.Iterate("", "", func(key string, value interface{}) bool {
        sb.WriteString("- ")
        sb.WriteString(value.(MessageHandler).Type())
        sb.WriteString("\n")
        return false
    })

    return sb.String()
}
```

File: examples/gno.land/p/demo/teritori/dao_proposal_single/dao_proposal_single.gno
Comment (notJoon) [inline examples/gno.land/p/demo/teritori/dao_proposal_single/dao_proposal_single.gno:147]: It seems we can use `ufmt.Sprintf` for formatting in this part.

```suggestion
		thresholdStr = ufmt.Sprintf("Threshold: %s\n\nQuorum: %s\n\n", threshold.Threshold.String(), threshold.Quorum.String())
```

File: docs/how-to-guides/creating-dao.md
Comment (notJoon) [inline docs/how-to-guides/creating-dao.md:158]: It's not crucial, but what do you think about marking changes in the code snippet like diff? 

For example, line that are added or modifed would be marked with a `+` sign, and lines that are deleted would be marked with a `-` sign. The downside in this case would be that we cannot use syntax highlighting.

Example:

```diff
import (
    "gno.land/p/demo/teritori/dao_interfaces"
+    "gno.land/p/demo/teritori/dao_voting_group"
)

func init() {
+    var group *dao_voting_group.VotingGroup

    votingModuleFactory := func(core dao_interfaces.IDAOCore) {
+        group = dao_voting_group.NewVotingGroup()
    }
}
```

Comment (n0izn0iz) [reply docs/how-to-guides/creating-dao.md:158]: I prefer to keep syntax highlighting and pastability

**Review by leohhhn — COMMENTED** [review]
Thank you for your awesome contribution, and thank you for helping make the docs better! I left reviews on the tutorial, mainly about the flow of it and typos. I haven't finished the review fully - I still want to go through the full tutorial hands-on. I am submitting this to save progress on the review :)

**Comment by n0izn0iz** [comment]
Thanks for the review @leohhhn I'm addressing your comments

I have two remarks to spark discussion around guidelines:

- Adding periods everywhere is [considered a bad ux practice by google](https://m3.material.io/foundations/content-design/style-guide/grammar-and-punctuation#555b17c7-de48-4974-8a82-51106899dc0e)

- Tutorials should be playful instead of formal
  > The use of humor in the college classroom has been researched extensively (see Segrist & Hupp, 2015 who summarized 41 years of literature on humor in the college classroom) and has been shown to have many benefits for students (Banas, Dunbar, Rodriguez & Liu, 2011; Garner, 2006; Huss & Estep, 2016; Pollio, 2002), some of which are:
    An increase in learning.
    An increase in self-motivation.
    An increase in class attendance.
    An increase in test performance.
    An increase in divergent thinking.
    An increase of interest in learning.
    A reduction of anxiety and stress in dealing with difficult material.
    The creation of a positive social and emotional learning environment.
    The creation of a common psychological bond between students and faculty.

**Comment by elix1er** [comment]
impressive work, keep grinding

File: docs/how-to-guides/creating-dao.md
Comment (leohhhn) [inline docs/how-to-guides/creating-dao.md:9]:
```suggestion
This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the realm, please see the
```

Comment (leohhhn) [inline docs/how-to-guides/creating-dao.md:9]:
```suggestion
This guide will show you how to write a simple [**DAO**](https://en.wikipedia.org/wiki/Decentralized_autonomous_organization) [realm](../concepts/realms.md) in [Gno](../concepts/gno-language.md). For actually deploying the realm, please see the
```

File: docs/how-to-guides/creating-dao.md
Comment (leohhhn) [inline docs/how-to-guides/creating-dao.md:88]:
```suggestion
Proposal actions are encoded as objects implementing `ExecutableMessage` found under `gno.land/p/demo/dao_maker/dao_interfaces`.
```

File: docs/how-to-guides/creating-dao.md
Comment (leohhhn) [inline docs/how-to-guides/creating-dao.md:99]:
```suggestion
They are unmarshalled and executed by message handlers implementing `gno.land/p/demo/dao_maker/dao_interfaces.MessageHandler`.
```
Marshal/unmarshal is a more "go way" to say serialize/deserialize.

File: docs/how-to-guides/creating-dao.md
Comment (leohhhn) [inline docs/how-to-guides/creating-dao.md:111]:
```suggestion
In this section, we will showcase how to implement your own DAO using the `dao_maker` package suite.

```

File: docs/how-to-guides/creating-dao.md
Comment (leohhhn) [inline docs/how-to-guides/creating-dao.md:121]:
```suggestion
- Initialize the gno module: `gno mod init gno.land/r/<your_namespace>/my_dao`
```

**Review by leohhhn — APPROVED** [review]
Left some minor suggestions to the docs page. Otherwise we should be good to go.

Sorry for the delay, and great work.

File: examples/gno.land/p/demo/users/types.gno
Comment (moul) [inline examples/gno.land/p/demo/users/types.gno:20]: what about using `users.Resolve(aon).IsValid()`?

File: examples/gno.land/p/demo/dao_maker/havl/havl.gno
Comment (moul) [inline examples/gno.land/p/demo/dao_maker/havl/havl.gno:1]: can you add more info about what is this package about?

it looks like something that should be located in `p/demo/havl`.

+ add some tests.

File: examples/gno.land/p/demo/dao_maker/jsonutil/gno.mod
Comment (moul) [inline examples/gno.land/p/demo/dao_maker/jsonutil/gno.mod:1]: The current `p/demo/json` library should be suitable for your needs.

I'm against having any JSON library written in userland because I expect that we can use @gfanton's #1776 and avoid any boilerplate code that isn't either for humans (rendering) or for other contracts (imports). Generally, anything about marshalling doesn't fall into these categories.

In the meantime, you can reuse the existing JSON library if needed. However, as soon as possible, I expect we can delete it completely. Additionally, if possible, I expect we don't merge any new JSON library in the meantime.

File: examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno
Comment (moul) [inline examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno:156]:
```suggestion
	sb.WriteString(ufmt.Sprintf("%d", havl.Latest))
```

File: examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno
Comment (moul) [inline examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno:141]: You can use `avl.Tree.IterateByOffset`.

File: examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno
Comment (moul) [inline examples/gno.land/p/demo/dao_maker/dao_voting_group/voting_group.gno:128]: isn't `avl.Tree.Size` enough for you?

File: examples/gno.land/r/demo/dao_maker/tori/tori.gno
Comment (moul) [inline examples/gno.land/r/demo/dao_maker/tori/tori.gno:1]: can you move this to `gno.land/r/teritori/tori`?

File: examples/gno.land/p/demo/dao_maker/dao_interfaces/messages.gno
Comment (moul) [inline examples/gno.land/p/demo/dao_maker/dao_interfaces/messages.gno:13]: why is this needed?

**Review by moul — CHANGES_REQUESTED** [review]
There are several interesting aspects in that PR, but there is too much work to review at the moment. Therefore, I will mark it as "request changes" to allow more time for a thorough review.

A high-level suggestion I can provide is to:
- Simplify the codebase by reducing the number of packages, files, functions, types, and boilerplate, and avoiding the use of JSON, etc. Consider merging some packages together.
- Improve the naming conventions. For instance, use more specific names like "Teritori Realms" and "Daomaker" (a better name could facilitate faster merging, as "Daomaker" is currently too generic and sounds overly official, while "Teritori" would make it easier to integrate).

File: examples/gno.land/p/demo/dao_maker/jsonutil/gno.mod
Comment (n0izn0iz) [reply examples/gno.land/p/demo/dao_maker/jsonutil/gno.mod:1]: this extends `p/demo/json`

File: examples/gno.land/p/demo/users/types.gno
Comment (n0izn0iz) [reply examples/gno.land/p/demo/users/types.gno:20]: can't call the realm method `users.Resolve` from a `p` package

**Comment by moul** [comment]
Can you move all the packages under `{r,p}/teritori/...` and relocate the `docs/...` into `r/teritori/dao_maker/README.md`?

I believe we should merge your PR in its current state but under teritori.

**Comment by Gno2D2** [comment]
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🔴 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🔴 The pull request head branch must be up-to-date with its base ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch))
 🔴 Changes to 'docs' folder must be reviewed/authored by at least one devrel and one tech-staff


## Manual Checks

*No manual checks match this pull request.*

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Head branch (TERITORI:daosdk) is up to date with base (master): behind by 490 / ahead by 111

```
</blockquote></details>

<details><summary>Changes to 'docs' folder must be reviewed/authored by at least one devrel and one tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 A changed file matches this pattern: ^docs/ (filename: docs/how-to-guides/creating-dao.md)

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Or
    ├── 🔴 And
    │   ├── 🔴 Pull request author is a member of the team: devrels
    │   └── 🔴 At least 1 user(s) of the team tech-staff approved pull request
    └── 🔴 And
        ├── 🔴 Pull request author is a member of the team: tech-staff
        └── 🟢 At least 1 user(s) of the team devrels approved pull request

```
</blockquote></details>

</blockquote></details>



</blockquote></details>

**Comment by n0izn0iz** [comment]
closing in favor of #3358 
will address reviews there

**Pull request closed without merging**
